src/codec.rs

```diff
@@ -1207,13 +1274,20 @@ macro_rules! impl_codec_through_iterator {
 			fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
 				<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
 					input.descend_ref()?;
-					let result = Result::from_iter((0..len).map(|_| Decode::decode(input)));
+					let result = Result::from_iter((0..len).map(|_| {
+						input.on_before_alloc_mem(0 $( + mem::size_of::<$generics>() )*)?;
```

Comment:
This is an underestimate of the allocation, is it fine?

All the collections below will use more space than what is reported.

maybe it should be documented in `Input` that it is a rough estimation then.

Reply:
Yes, this is as desired. It is acceptable to be a bit off. The main thing is to track the memory used by the decoded data inside the type.

You're right, it's good to document this. Added a comment in `Input`.

Reply:
great :+1: